src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,210 @@
+
+def hamm_dist (x y : Π i, β i) := card {i // x i ≠ y i}
+
+lemma hamm_dist_smul_le [Π i, has_scalar α (β i)] (k : α) (x y : Π i, β i) :
+hamm_dist (k • x) (k • y) ≤ hamm_dist x y :=
+card_subtype_mono _ _ (λ i h H, h (by rw [pi.smul_apply, pi.smul_apply, H]))
+
+lemma hamm_dist_smul [Π i, has_scalar α (β i)] {k : α}
+(hk : ∀ i, is_smul_regular (β i) k) (x y : Π i, β i) :
+hamm_dist x y = hamm_dist (k • x) (k • y) :=
+le_antisymm (card_subtype_mono _ _ (λ _ h H, h (hk _ H))) (hamm_dist_smul_le _ _ _)
+
+lemma hamm_dist_eq (x y : Π i, β i) : hamm_dist x y = card {i // x i ≠ y i} := rfl
+
+lemma hamm_dist_comm (x y : Π i, β i) : hamm_dist x y = hamm_dist y x :=
+by simp_rw [hamm_dist_eq, ne_comm]
+
+lemma hamm_dist_triangle (x y z : Π i, β i) : hamm_dist x z ≤ hamm_dist x y + hamm_dist y z :=
+begin
+  simp_rw hamm_dist_eq,
+  refine le_trans (card_subtype_mono _ _ (λ _ h, _)) (card_subtype_or _ _),
+  by_contra' H, exact h (eq.trans H.1 H.2)
+end
+
+lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
+begin
+  rw [function.funext_iff, hamm_dist_eq, card_eq_zero_iff],
+  exact ⟨ λ h i, imp_of_not_imp_not _ _ (λ H, h.elim' ⟨i, H⟩) h,
+          λ h, subtype.is_empty_of_false (λ i H, H (h _))⟩
+end
+
+@[simp] lemma hamm_dist_self (x : Π i, β i) : hamm_dist x x = 0 := hamm_dist_eq_zero.mpr rfl
+
+lemma eq_of_hamm_dist_eq_zero (x y : Π i, β i) :
+hamm_dist x y = 0 → x = y := hamm_dist_eq_zero.mp
+
+lemma hamm_dist_ne_zero {x y : Π i, β i} : hamm_dist x y ≠ 0 ↔ x ≠ y :=
+not_iff_not.mpr hamm_dist_eq_zero
+
+lemma hamm_dist_pos {x y : Π i, β i} : 0 < hamm_dist x y ↔ x ≠ y :=
+by rw [←hamm_dist_ne_zero, iff_not_comm, not_lt, nat.le_zero_iff]
+
+@[simp] lemma hamm_dist_eq_zero_iff_forall_eq {x y : Π i, β i} :
+hamm_dist x y = 0 ↔ ∀ i, x i = y i := by rw [hamm_dist_eq_zero, function.funext_iff]
+
+lemma hamm_dist_ne_zero_iff_exists_ne {x y : Π i, β i} :
+hamm_dist x y ≠ 0 ↔ ∃ i, x i ≠ y i := by rw [hamm_dist_ne_zero, function.ne_iff]
+
+section has_zero
+
+variable [Π i, has_zero (β i)]
+
+/-- The Hamming weight function to the naturals. -/
+
+def hamm_wt (x : Π i, β i) : ℕ := hamm_dist x 0
+
+lemma hamm_wt_eq (x : Π i, β i) : hamm_wt x = card {i // x i ≠ 0} := rfl
+
+lemma hamm_wt_eq_hamm_dist_zero (x : Π i, β i) : hamm_wt x = hamm_dist x 0 := rfl
+
+lemma hamm_wt_smul_le [has_zero α] [Π i, smul_with_zero α (β i)] (k : α)
+(x : Π i, β i) : hamm_wt (k • x) ≤ hamm_wt x :=
+by convert hamm_dist_smul_le k x _; rw smul_zero'; refl
+
+lemma hamm_wt_smul [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+(hk : ∀ i, is_smul_regular (β i) k) (x : Π i, β i) : hamm_wt x = hamm_wt (k • x) :=
+by convert hamm_dist_smul hk _ _; rw smul_zero'; refl
+
+lemma hamm_wt_eq_zero {x : Π i, β i} : hamm_wt x = 0 ↔ x = 0 := hamm_dist_eq_zero
+
+@[simp] lemma hamm_wt_zero : hamm_wt (0 : Π i, β i) = 0 := hamm_dist_self _
+
+lemma zero_of_hamm_wt_eq_zero (x : Π i, β i) :
+hamm_wt x = 0 → x = 0 := eq_of_hamm_dist_eq_zero _ _
```

Comment:
This lemma doesn't seem that useful in the presence of `hamm_wt_eq_zero`

Reply:
Similarly I think I can get rid of `eq_of_hamm_dist_eq_zero`.